### PR TITLE
Merging work on reward shaping and annealing 

### PR DESCRIPTION
**Pull request by kantneel**
- Added reward_shaping.py which has class for a RewardShapingEnv, a Scheduler, and a few Annealers. 
- Modified ppo_baseline to support experiments where reward shaping is annealed over a fraction of training. Also data is logged in tensorboard format in addition to stdout.
- Added default options to ppo_baseline and score_agent for human sumo experiments

### CODE REVIEW

**Comment by AdamGleave** [comment]
I've made a lot of changes to `master`, which should simplify implementation of some of these things (particularly adding noise to the victim), so now would be a good time to merge `master` again. Ping me once you've done that & got the unit tests passing and I'll do a more detailed review. In the meantime will go through and leave some high-level comments.

File: src/modelfree/ppo_and_score.py [hunk 1]
```diff
@@ -16,6 +16,18 @@ def default_config():
     }
 
 
+@ppo_and_score_ex.named_config
+def ant_score_config():
+    config = {  # noqa: F841
+        "ppo": {"victim": "/home/neel/multiagent-competition/agent-zoo/" + \
```
Comment (AdamGleave) [inline src/modelfree/ppo_and_score.py]: Don't hardcode paths. This also seems like an outdated way of specifying zoo param locations.

File: src/modelfree/ppo_baseline.py [hunk 2]
```diff
@@ -22,6 +22,9 @@
 
 from aprl.envs.multi_agent import CurryEnv, FlattenSingletonEnv
 from modelfree.gym_compete_conversion import GymCompeteToOurs, load_zoo_agent
+from modelfree.reward_shaping import (LinearAnnealer, RewardShapingEnv, Scheduler,
```
Comment (AdamGleave) [inline src/modelfree/ppo_baseline.py]: Classes should go first, and be in alphabetical order, then functions. isort will do this for you automatically, please run ci/code_checks.sh

File: src/modelfree/ppo_baseline.py [hunk 3]
```diff
@@ -89,13 +92,33 @@ def make_single_env(env_name, seed, agent_fn, out_dir, env_id=0):
     return single_env
 
 
+def make_real_single_env(env_name, seed, out_dir, env_id=0):
```
Comment (AdamGleave) [inline src/modelfree/ppo_baseline.py]: Seems like a lot of code duplication with make_single_env, try to refactor.

File: src/modelfree/ppo_baseline.py [hunk 4]
```diff
@@ -206,22 +247,45 @@ def setup_logger(out_dir="results", exp_name="test"):
...
+    logger.configure(dir=osp.join(out_dir, 'mon'),
+                     format_strs=['tensorboard', 'stdout'])
     return out_dir
 
 
 ppo_baseline_ex = Experiment("ppo_baseline")
 ppo_baseline_ex.observers.append(FileStorageObserver.create('my_runs'))
 
 
+@ppo_baseline_ex.named_config
```
Comment (AdamGleave) [inline src/modelfree/ppo_baseline.py]: Can some of these be inherited from `default_ppo_config` and so do not need to be specified again?

Conversely, this defines some params (`env_wrapper_type`, ...) that do not appear in `default_ppo_config`. Should they defined in `default_ppo_config` too?

File: src/modelfree/ppo_baseline.py [hunk 5]
```diff
@@ -231,21 +295,60 @@ def default_ppo_config():
...
+    scheduler = Scheduler(lr_func=annealer_collection['default_lr'].get_value)
+
+    if rew_shape_anneal_frac is not None:
+        assert 0 <= rew_shape_anneal_frac <= 1
+        rew_shape_func = LinearAnnealer(1, 0, rew_shape_anneal_frac).get_value
+        scheduler.set_rew_shape_func(rew_shape_func)
+
+        def env_wrapper(x):
+            return RewardShapingEnv(x, reward_annealer=scheduler.get_rew_shape_val)
+    else:
```
Comment (AdamGleave) [inline src/modelfree/ppo_baseline.py]: Should this really be an else? Might we want to do e.g. both HumanoidEnvWrapper and RewardShapingEnv?

File: src/modelfree/reward_shaping.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    default_shaping_params = {
+        # 'center_reward' : 1,
+        # 'ctrl_cost'     : -1,
+        # 'contact_cost'  : -1,
+        'survive': 10,
+
+        # sparse reward field as per gym_compete/new_envs/multi_agent_env:151
+        'reward_remaining': 1,
+        # dense reward field as per gym_compete/new_envs/agents/humanoid_fighter:45
+        'reward_move': 10
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: Suggest leaving trailing commas at end of dicts/lists see https://stackoverflow.com/questions/11597901/why-are-trailing-commas-allowed-in-a-list

File: src/modelfree/reward_shaping.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+import numpy as np
+import logging
+import time
+from baselines.common.vec_env import VecEnvWrapper
+from baselines.bench.monitor import ResultsWriter
+from modelfree.simulation_utils import ResettableAgent
+from gym.envs.mujoco import HumanoidEnv
+
+
+class RewardShapingEnv(VecEnvWrapper):
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: We've already discussed verbally some changes to this, deferring review until after it's been updated.

File: src/modelfree/reward_shaping.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+        noise = self.noise_generator(noise_param, action_size)
+        noisy_action = original_action + noise
+        return noisy_action
+
+    def reset(self):
+        return self.agent.reset()
+
+
+class HumanoidEnvWrapper(VecEnvWrapper):
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: Why can't you use RewardShapingEnv for this?

File: src/modelfree/reward_shaping.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    def step_wait(self):
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
+
+        return obs, rew, done, infos
+
+
+# TODO: should this inherit from something?
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: Probably not, no.

File: src/modelfree/reward_shaping.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+class Scheduler(object):
+    """Keep track of frac_remaining and return time-dependent values"""
+    schedule_num = 0
+
+    def __init__(self, lr_func=None, rew_shape_func=None, noise_func=None):
+        # print('made a scheduler')
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: remove debugging statements

File: src/modelfree/reward_shaping.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
+
+        return obs, rew, done, infos
+
+
+# TODO: should this inherit from something?
+class Scheduler(object):
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: There's a lot of code duplication here. You're doing basically the same thing for {get,set}_{lr,rew,noise}_{func,val}. Can you refactor this to have a common interface?

I'm thinking take a dict of callables, and then just have a get_val(k, frac_remaining) function which: (a) updates frac remaining if non-None, (b) calls func[k](self.frac_remaining). You can also have a `set_func(k, f)`, although I'm not sure whether this is used.

File: src/modelfree/reward_shaping.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+    def __init__(self, lr_func=None, rew_shape_func=None, noise_func=None):
+        # print('made a scheduler')
+        # print(Scheduler.schedule_num)
+        self._lr_func = lr_func
+        self._rew_shape_func = rew_shape_func
+        self._noise_func = noise_func
+
+        self.frac_remaining = 1  # frac_remaining goes from 1 to 0
+        self.same_thing = False
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: debugging statement. suggest adding `# DEBUG` or `# TODO: remove` or similar so you remember to kill these.

File: src/modelfree/reward_shaping.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        Scheduler.schedule_num += 1
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: debugging statement

File: src/modelfree/reward_shaping.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
+
+        return obs, rew, done, infos
+
+
+# TODO: should this inherit from something?
+class Scheduler(object):
+    """Keep track of frac_remaining and return time-dependent values"""
+    schedule_num = 0
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: debugging statement

File: src/modelfree/reward_shaping.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+# Annealers
+
+class Annealer(ABC):
+    """Abstract class for implementing Annealers."""
+    def __init__(self, start_val, end_val):
+        self.start_val = start_val
+        self.end_val = end_val
+
+    @abstractmethod
+    def get_value(self, step):
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: s/step/frac_remaining

File: src/modelfree/reward_shaping.py [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+
+    def get_noise_val(self, frac_remaining=None):
+        self._update_frac_remaining(frac_remaining)
+        assert callable(self._noise_func)
+        return self._noise_func(self.frac_remaining)
+
+
+# Annealers
+
+class Annealer(ABC):
```
Comment (AdamGleave) [inline src/modelfree/reward_shaping.py]: yes, nice use of a base class :)

File: src/modelfree/score_agent.py [hunk 17]
```diff
@@ -45,6 +45,21 @@ def get_agent_any_type(agent, agent_type, env, env_name, index, sess=None):
 score_agent_ex.observers.append(FileStorageObserver.create('my_runs'))
 
 
+@score_agent_ex.named_config
```
Comment (AdamGleave) [inline src/modelfree/score_agent.py]: some parameters duplicated with default_score_config

File: src/modelfree/simulation_utils.py [hunk 18]
```diff
@@ -20,6 +20,23 @@ def simulate(env, agents, render=False):
 
         yield observations, rewards, dones, infos
 
+def simulate_single(env, agent, render=False):
```
Comment (AdamGleave) [inline src/modelfree/simulation_utils.py]: Do we really need a function for both multi-agent and single-agent simulation? :(

Perhaps make a wrapper that turns a (Vec)Env into a (Vec)MultiEnv with a single agent?

File: src/modelfree/single_score_agent.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
+import gym
```
Comment (AdamGleave) [inline src/modelfree/single_score_agent.py]: No! :( This can definitely be  supported within score_agent. Please avoid this level of code duplication. @MichaelDDennis

File: src/modelfree/ppo_baseline.py [hunk 20]
```diff
@@ -91,23 +104,39 @@ def default_ppo_config():
     batch_size = 2048               # batch size
     seed = 0
     load_path = None                # path to load initial policy from
+    env_wrapper_type = None
```
Comment (AdamGleave) [inline src/modelfree/ppo_baseline.py]: Add comments? (These will be included in --help so are quite useful)

File: src/modelfree/scheduling.py [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    @abstractmethod
+    def get_value(self, step):
+        raise NotImplementedError()
+
+
+class ConstantAnnealer(Annealer):
+    """Returns a constant value"""
+    def __init__(self, const_val):
+        self.const_val = const_val
+        super().__init__(const_val, None)
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/scheduling.py]: If you only ever use the `get_value`, it might be cleaner to implement this as a decorator, although I don't have strong feelings here.

File: src/modelfree/shaping_wrappers.py [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+        super().__init__(env)
+        self.shaping_params = shaping_params
+        self.reward_annealer = reward_annealer
+        self.counter = 0
+
+        self.ep_rew_dict = defaultdict(list)
+        self.step_rew_dict = defaultdict(lambda: [[] for _ in range(self.num_envs)])
+
+    def _log_sparse_dense_rewards(self):
+        if self.counter == 2048 * self.num_envs:
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: The semantics of logging every `2048 * self.num_envs` timesteps seem odd to me. Main issues are:

  * No guarantee that this lines up with episode boundaries, so we can have variable number of episodes in our mean.
  * This number of timesteps does not correspond to anything else in the algorithm, so it's another number the user has to remember. If we wanted to log after a fixed number of timesteps, the batch size would be a natural metric.
  * There's no way to change the logging period.
  * What is `2048` anyway (avoid magic numbers).

File: src/modelfree/shaping_wrappers.py [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+
+        # sparse reward field as per gym_compete/new_envs/multi_agent_env:151
+        'reward_remaining': 1,
+        # dense reward field as per gym_compete/new_envs/agents/humanoid_fighter:45
+        # 'reward_move': 10,
+    }
+
+    def __init__(self, env, shaping_params=default_shaping_params,
+                 reward_annealer=None):
+        super().__init__(env)
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: s/env/venv/

File: src/modelfree/shaping_wrappers.py [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+        self.reward_annealer = reward_annealer
+        self.counter = 0
+
+        self.ep_rew_dict = defaultdict(list)
+        self.step_rew_dict = defaultdict(lambda: [[] for _ in range(self.num_envs)])
+
+    def _log_sparse_dense_rewards(self):
+        if self.counter == 2048 * self.num_envs:
+            num_episodes = len(self.ep_rew_dict['reward_remaining'])
+            dense_terms = ['reward_center', 'reward_ctrl', 'reward_contact', 'reward_survive']
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Don't like that dense terms are hardcoded. Perhaps we can take as input to `__init__` a sparse (non-annealed) dict of weights, and a dense (annealed) dict of weights, then extract the keys from there?

File: src/modelfree/shaping_wrappers.py [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+            for term in dense_terms:
+                assert len(self.ep_rew_dict[term]) == num_episodes
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: this is a good check :+1

File: src/modelfree/shaping_wrappers.py [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+            self.counter = 0
+
+    def reset(self):
+        return self.venv.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.venv.step_wait()
+
+        # replace rew with differently shaped rew
+        # victim is agent 0, attacker is agent 1
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Can you make the attacker agent index a parameter to `__init__`? Right now all our experiments have it be index `1`, but this could change in the future (e.g. maybe we want to attack the goalie in `KickAndDefend-v0` rather than the shooter).

File: src/modelfree/shaping_wrappers.py [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+            shaped_reward = 0
+            for rew_type, rew_value in infos[env_num][1].items():
+                if rew_type not in self.shaping_params:
+                    continue
+
+                weighted_reward = self.shaping_params[rew_type] * rew_value
+                self.step_rew_dict[rew_type][env_num].append(weighted_reward)
+                if self.reward_annealer is not None:
+                    c = self.get_annealed_exploration_reward()
+                    if rew_type == 'reward_remaining':
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Also not a fan of the sparse reward term being hardcoded, although this seems less bad (since it really will be the same across all the `gym_compete` environments).

File: src/modelfree/shaping_wrappers.py [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+        # victim is agent 0, attacker is agent 1
+        for env_num in range(self.num_envs):
+            shaped_reward = 0
+            for rew_type, rew_value in infos[env_num][1].items():
+                if rew_type not in self.shaping_params:
+                    continue
+
+                weighted_reward = self.shaping_params[rew_type] * rew_value
+                self.step_rew_dict[rew_type][env_num].append(weighted_reward)
+                if self.reward_annealer is not None:
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: What happens if `self.reward_annealer is None`? `shaped_reward` is always zero?

File: src/modelfree/shaping_wrappers.py [hunk 29]
```diff
@@ -0,0 +1,208 @@
...
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: This is fine and good-practice to sanity check values, but possible alternative idea: have `get_value` in `Annealer` base class call `_get_value` (defined in subclasses) and do this sanity checking. That way all end-users benefit from this.

You could also add test cases for your annealers that ensure they always have the right range.

File: src/modelfree/shaping_wrappers.py [hunk 30]
```diff
@@ -0,0 +1,208 @@
...
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(BaseRLModel):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        Wrap an agent and add noise to its actions
+        :param agent: PolicyToModel(BaseRLModel) the agent to wrap
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Do you depend on it being `PolicyToModel`? If not, the type is just `BaseRLModel`

File: src/modelfree/shaping_wrappers.py [hunk 31]
```diff
@@ -0,0 +1,208 @@
...
+        over time in order to get the adversarial policy to perform well on a normal victim.
+        :param noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.sess = agent.sess
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    def _get_noise_generator(self, noise_type):
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: :+1:

File: src/modelfree/shaping_wrappers.py [hunk 32]
```diff
@@ -0,0 +1,208 @@
...
+
+    def predict(self, observation, state=None, mask=None, deterministic=False):
+        original_actions, states = self.agent.predict(observation, state, mask, deterministic)
+        action_shape = original_actions.shape
+
+        noise_param = self.noise_annealer()
+        noise = self.noise_generator(noise_param, action_shape)
+        normed_noise = original_actions * noise
+
+        noisy_actions = original_actions + normed_noise
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Or `original_actions * (1 + noise)`?

It's not clear to me whether additive or multiplicative noise will work best; let's discuss in person.

File: src/modelfree/shaping_wrappers.py [hunk 33]
```diff
@@ -0,0 +1,208 @@
...
+        raise NotImplementedError()
+
+    def save(self, save_path):
+        raise NotImplementedError()
+
+    def load(self):
+        raise NotImplementedError()
+
+
+class HumanoidEnvWrapper(VecEnvWrapper):
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Confused by what this does. Turn a walking version of Humanoid into a standing one? Should add a docstring.

Also, should this really be a wrapper? Maybe this should be a new Gym environment? Internally, the Gym environment can create `Humanoid` inside itself and then patch up its reward.

File: src/modelfree/shaping_wrappers.py [hunk 34]
```diff
@@ -0,0 +1,208 @@
...
+        self.env = env
+
+    def reset(self):
+        return self.env.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Can we not use `RewardShapingEnv` for this? I guess that assumes the info dicts have an axis corresponding to number of agents, but doesn't seem too hard to generalize to handle both? (Can have `agent_idx` be None (single-agent info dict), `0` or `1`.)

File: src/modelfree/shaping_wrappers.py [hunk 35]
```diff
@@ -0,0 +1,208 @@
...
+import numpy as np
+from stable_baselines import logger
+from stable_baselines.common.base_class import BaseRLModel
+from stable_baselines.common.vec_env import VecEnvWrapper
+
+from modelfree import envs  # noqa - needed to register sumo_auto_contact
+from modelfree.scheduling import LinearAnnealer
+
+
+class RewardShapingEnv(VecEnvWrapper):
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Consider renaming to `RewardShapingVecWrapper`: it's not an `Env`, it's a `VecEnvWrapper`.

File: src/modelfree/shaping_wrappers.py [hunk 36]
```diff
@@ -0,0 +1,208 @@
...
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
+
+        return obs, rew, done, infos
+
+
+def apply_env_wrapper(single_env, rew_shape_params, env_wrapper_type, scheduler):
+    if rew_shape_params is not None:
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: This looks weird.

File: src/modelfree/shaping_wrappers.py [hunk 37]
```diff
@@ -0,0 +1,208 @@
...
+            rew_shape_func = LinearAnnealer(1, 0, rew_shape_anneal_frac).get_value
+        else:
+            rew_shape_func = None
+
+        scheduler.set_func('rew_shape', rew_shape_func)
+        return RewardShapingEnv(single_env, shaping_params=shaping_params,
+                                reward_annealer=scheduler.get_func('rew_shape'))
+    else:
+        wrappers = {
+            'humanoid': HumanoidEnvWrapper,
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Might be unnecessary if we just make this a new environment (see earlier comment).

File: src/modelfree/shaping_wrappers.py [hunk 38]
```diff
@@ -0,0 +1,208 @@
...
+
+def apply_victim_wrapper(victim, victim_noise_anneal_frac, victim_noise_param, scheduler):
+    if victim_noise_anneal_frac is not None:
+        assert 0 <= victim_noise_anneal_frac <= 1
+        assert 0 < victim_noise_param
+        noise_anneal_func = LinearAnnealer(victim_noise_param, 0,
+                                           victim_noise_anneal_frac).get_value
+        scheduler.set_func('noise', noise_anneal_func)
+        return NoisyAgentWrapper(victim, noise_annealer=scheduler.get_func('noise'))
+
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: unnecessary whitespace

File: src/modelfree/simulation_utils.py [hunk 39]
```diff
@@ -0,0 +1,61 @@
+def simulate(env, agents, render=False):
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/simulation_utils.py]: I removed this in a commit.

File: src/modelfree/scheduling.py [hunk 40]
```diff
@@ -0,0 +1,93 @@
+from abc import ABC, abstractmethod
+import functools
+
+
+class Scheduler(object):
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/scheduling.py]: Thanks!

File: src/modelfree/gym_compete_conversion.py [hunk 41]
```diff
@@ -42,7 +42,7 @@ def get_policy_type_for_agent_zoo(env_name):
         "RunToGoalAnts-v0": "mlp",
         "YouShallNotPassHumans-v0": "mlp",
         "SumoHumans-v0": "lstm",
-        "SumoAnts-v0": "lstm"
+        "SumoAnts-v0": "lstm",
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/gym_compete_conversion.py]: Haha

File: src/modelfree/shaping_wrappers.py (hunk 25 above)
Comment (kantneel) [reply src/modelfree/shaping_wrappers.py]: Thanks

File: tests/modelfree/test_experiments.py [hunk 42]
```diff
@@ -48,6 +49,11 @@ def test_score_agent(config):
 
 
 PPO_BASELINE_CONFIGS = [
+    {
```
[the remark this answers is not in the document]
Comment (kantneel) [reply tests/modelfree/test_experiments.py]: yep

File: src/modelfree/shaping_wrappers.py [hunk 43]
```diff
@@ -0,0 +1,208 @@
...
+        'reward_contact': -1,
+        'reward_survive': 1,
+
+        # sparse reward field as per gym_compete/new_envs/multi_agent_env:151
+        'reward_remaining': 1,
+        # dense reward field as per gym_compete/new_envs/agents/humanoid_fighter:45
+        # 'reward_move': 10,
+    }
+
+    def __init__(self, env, shaping_params=default_shaping_params,
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/shaping_wrappers.py]: nice I like that.

File: src/modelfree/score_agent.py [hunk 44]
```diff
@@ -52,6 +51,13 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
 score_agent_ex.observers.append(FileStorageObserver.create('data/sacred'))
 
 
+@score_agent_ex.named_config
+def human_score_config():
+    env_name = "multicomp/SumoHumans-v0"
+    _ = locals()  # quieten flake8 unused variable warning
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/score_agent.py]: okay will do

File: src/modelfree/rew_configs/anneal_0.1.json [hunk 45]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/rew_configs/anneal_0.1.json]: sounds good.

File: src/modelfree/gym_compete_conversion.py [hunk 46]
```diff
@@ -131,6 +131,11 @@ def load_zoo_policy(tag, policy_type, scope, env, env_name, index):
 
 
 def load_zoo_agent(path, env, env_name, index):
+    symlink_dict = {
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/gym_compete_conversion.py]: sure

File: src/modelfree/shaping_wrappers.py (hunk 35 above)
Comment (kantneel) [reply src/modelfree/shaping_wrappers.py]: sure

File: src/modelfree/shaping_wrappers.py (hunk 28 above)
Comment (kantneel) [reply src/modelfree/shaping_wrappers.py]: No. shaped_reward is accumulated at line 84 regardless. If `self.reward_annealer is None` then there will not be any annealing of dense reward to zero.

File: src/modelfree/shaping_wrappers.py (hunk 29 above)
Comment (kantneel) [reply src/modelfree/shaping_wrappers.py]: Well there are plenty of Annealers which need not be bounded in [0, 1] e.g. learning rate.

File: src/modelfree/ppo_baseline.py (hunk 4 above)
Comment (kantneel) [reply src/modelfree/ppo_baseline.py]: Yes this has been taken care of.

File: src/modelfree/shaping_wrappers.py (hunk 32 above)
Comment (kantneel) [reply src/modelfree/shaping_wrappers.py]: Yes that is clearer.

File: src/modelfree/shaping_wrappers.py (hunk 28 above)
Comment (AdamGleave) [reply src/modelfree/shaping_wrappers.py]: you're right.

File: src/modelfree/shaping_wrappers.py (hunk 29 above)
Comment (AdamGleave) [reply src/modelfree/shaping_wrappers.py]: fair point.

File: src/modelfree/shaping_wrappers.py [hunk 47]
```diff
@@ -0,0 +1,215 @@
...
+    def get_shaping_params(cls, shaping_params, env_name):
+        path_stem = 'experiments/rew_configs/'
+        default_configs = {
+            'multicomp/SumoHumans-v0': 'default_hsumo.json',
+            'multicomp/SumoHumansAutoContact-v0': 'default_hsumo.json'
+        }
+        assert env_name in default_configs, \
+            "No default reward shaping config found for env_name:{}".format(env_name)
+        fname = os.path.join(path_stem, default_configs[env_name])
+        final_shaping_params = json.load(open(fname))
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: This will leak file descriptors, it should be:
```
with open(fname) as f:
    final_shaping_params = json.load(f)
```

This is also fragile since `fname` depends on your current directory. I suggest instead adding the configs to `setup.py` using the `data_files` keyword, see https://docs.python.org/3/distutils/setupscript.html#installing-additional-files, and then using `pkgresources` to load it.

File: src/modelfree/shaping_wrappers.py [hunk 48]
```diff
@@ -0,0 +1,215 @@
...
+
+        if shaping_params is not None:
+            config = json.load(open(shaping_params))
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Same comment as above re: leaking file descriptors.

File: src/modelfree/shaping_wrappers.py [hunk 49]
```diff
@@ -0,0 +1,215 @@
...
+            final_shaping_params['sparse'].update(config['sparse'])
+            final_shaping_params['dense'].update(config['dense'])
+            final_shaping_params['rew_shape_anneal_frac'] = config.get('rew_shape_anneal_frac')
+        return final_shaping_params
+
+    def _log_sparse_dense_rewards(self):
+        if self.counter == self.batch_size * self.num_envs:
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: This is better than before but really shouldn't depend on `batch_size` at all :( Can you do the callback approach instead?

File: src/modelfree/shaping_wrappers.py [hunk 50]
```diff
@@ -0,0 +1,215 @@
...
+        :param noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.sess = agent.sess
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    @classmethod
+    def get_noise_params(cls, noise_params, env_name):
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Deja-vu with `get_shaping_params`. Refactor this `json` loading code out to a separate function to code down on code duplication.

File: tests/modelfree/test_experiments.py [hunk 51]
```diff
@@ -51,6 +52,14 @@ def test_score_agent(config):
     {'num_env': 1},
     {'env_name': 'multicomp/KickAndDefend-v0'},
     {'victim_type': 'ppo2', 'victim_path': os.path.join(BASE_DIR, 'dummy_sumo_ants.pkl')},
+    {
+        'rew_shape_params': 'experiments/rew_configs/densex10_0.1.json',
+        'env_name': 'multicomp/SumoHumans-v0'
+    },
+    {
+        'victim_noise_params': 'experiments/noise_configs/frac0.5_noise0.5.json',
```
Comment (AdamGleave) [inline tests/modelfree/test_experiments.py]: :+1:

File: src/modelfree/shaping_wrappers.py [hunk 52]
```diff
@@ -7,66 +7,52 @@
...
-            logger.logkv('rew_anneal', c)
-            ep_rew_mean = c * ep_dense_mean + (1 - c) * ep_sparse_mean
-            logger.logkv('eprewmean_true', ep_rew_mean)
-
-            for rew_type in self.ep_rew_dict:
-                self.ep_rew_dict[rew_type] = []
-            self.counter = 0
+    def log_sparse_dense_rewards(self, locals, globals):
+        """Logs various metrics. This is given as a callback to PPO2.learn()"""
+        _ = [locals, globals]
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: You could also have this not take `locals, globals` and just have a `lambda` function swallow those arguments when you add it to the callback: `lambda locals, globals: single_env.log_sparse_dense_rewards()`. This seems cleaner unless it might need to know about `locals` or `globals` in the future.

File: src/modelfree/shaping_wrappers.py [hunk 53]
```diff
@@ -186,8 +153,30 @@ def load(self):
         raise NotImplementedError()
 
 
+def load_wrapper_params(params_path, env_name, rew_shaping=False, noisy_victim=False):
+    config_dir = 'rew_configs' if rew_shaping else 'noise_configs'
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Should either add an assertion that `rew_shaping` and `noisy_victim` are not both `True`, or just take in directly a `config_dir` directory and check it is either `rew_configs` or `noise_configs` (then change behaviour later based on which of these it is ).

File: src/modelfree/shaping_wrappers.py [hunk 54]
```diff
@@ -186,8 +153,30 @@ def load(self):
...
+        'multicomp/SumoHumansAutoContact-v0': 'default_hsumo.json'
+    }
+    fname = os.path.join(path_stem, default_configs[env_name])
+    with open(fname) as default_config_file:
+        final_params = json.load(default_config_file)
+    if params_path != 'default':
+        with open(params_path) as config_file:
+            config = json.load(config_file)
+        if rew_shaping:
+            final_params['sparse'].update(config.get('sparse', {}))
```
Comment (AdamGleave) [inline src/modelfree/shaping_wrappers.py]: Is the only reason you need to special case for `rew_shaping` because it has nested dicts? You could write a recursive implementation that merges possibly-nested `dict`'s. Probably not worth it right now, but would consider it if you have more configs in the future.
